lightning/restore/restore.go

```diff
@@ -304,48 +305,236 @@ outside:
-
-		// restore views. Since views can cross database we must restore views after all table schemas are restored.
-		for _, dbMeta := range rc.dbMetas {
-			if len(dbMeta.Views) > 0 {
-				task := log.With(zap.String("db", dbMeta.Name)).Begin(zap.InfoLevel, "restore view schema")
-				viewsSchema := make(map[string]string)
-				for _, viewMeta := range dbMeta.Views {
-					viewsSchema[viewMeta.Name] = viewMeta.GetSchema(ctx, rc.store)
+	}
+	err = worker.wait()
+	if err != nil {
+		return err
+	}
+	// 3. restore views. Since views can cross database we must restore views after all table schemas are restored.
+	for _, dbMeta := range dbMetas {
+		for _, viewMeta := range dbMeta.Views {
+			sql := viewMeta.GetSchema(worker.ctx, worker.store)
+			if sql != "" {
+				stmts, err := createTableIfNotExistsStmt(worker.glue.GetParser(), sql, dbMeta.Name, viewMeta.Name)
+				if err != nil {
+					return err
+				}
+				restoreSchemaJob := &schemaJob{
+					dbName:   dbMeta.Name,
+					tblName:  viewMeta.Name,
+					stmtType: schemaCreateView,
+					stmts:    make([]*schemaStmt, 0, len(stmts)),
+				}
+				for _, sql := range stmts {
+					restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+						sql: sql,
+					})
+				}
+				err = worker.appendJob(restoreSchemaJob)
+				if err != nil {
+					return err
 				}
-				err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, viewsSchema)
+				// we don't support restore views concurrency, cauz it maybe will raise a error
+				err = worker.wait()
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
 
+func (worker *restoreSchemaWorker) doJob() {
+	var session checkpoints.Session
+	defer func() {
+		if session != nil {
+			session.Close()
+		}
+	}()
+	for {
+		select {
+		case <-worker.ctx.Done():
+			// don't throw `worker.ctx.Err()` here,
+			// it will be blocked to death when number of `doJob`'s goroutines  > 1.
+			return
+		case job := <-worker.jobCh:
+			if job == nil {
+				return
+			}
+			var err error
+			if session == nil {
+				session, err = worker.glue.GetSession(worker.ctx)
+				if err != nil {
+					worker.wg.Done()
+					worker.throw(err)
+					return
+				}
+			}
+			logger := log.With(zap.String("db", job.dbName), zap.String("table", job.tblName))
+			for _, stmt := range job.stmts {
+				task := logger.Begin(zap.DebugLevel, fmt.Sprintf("execute SQL: %s", stmt.sql))
+				_, err = session.Execute(worker.ctx, stmt.sql)
 				task.End(zap.ErrorLevel, err)
 				if err != nil {
-					return errors.Annotatef(err, "restore view schema %s failed", dbMeta.Name)
+					err = errors.Annotatef(err, "%s %s failed", job.stmtType.String(), common.UniqueTable(job.dbName, job.tblName))
+					worker.wg.Done()
+					worker.throw(err)
+					return
 				}
 			}
+			worker.wg.Done()
+		}
+	}
+}
 
+func (worker *restoreSchemaWorker) wait() error {
+	select {
+	case err := <-worker.errCh:
+		return err
+	case <-worker.ctx.Done():
+		return worker.ctx.Err()
+	default:
+		worker.wg.Wait()
```

Comment:
if `wait()` goes here there's no chance to let it aware of parent context done, because `doJob` will exit on context done and don't call `wg.Done`.

Reply:
you may refer to 
https://github.com/pingcap/tidb-lightning/blob/b23840d3fc3c61a2517e2cf97ed03b5fc302faf7/lightning/restore/restore.go#L712-L729
https://github.com/pingcap/tidb-lightning/blob/b23840d3fc3c61a2517e2cf97ed03b5fc302faf7/lightning/restore/restore.go#L826-L846

to refine your concurrency model

Reply:
Yes, if `wg.Wait()` blocked but goroutines of `doJob` exit too early(cauz `ctx.Done` || `err`), it will wait forever. PR updated, PTAL! xD